Sources/SourceControl/GitRepository.swift

```diff
@@ -325,9 +325,9 @@ public class GitRepository: Repository, WorkingCheckout {
 
     public func hasUnpushedCommits() throws -> Bool {
         return try queue.sync {
-            let hasOutput = try Process.checkNonZeroExit(
-                args: Git.tool, "-C", path.asString, "log", "--branches", "--not", "--remotes").chomp().isEmpty
-            return !hasOutput
+            let output = try Process.checkNonZeroExit(
+                args: Git.tool, "-C", path.asString, "branch", "-r", "--contains", "HEAD").chomp()
```

Comment:
The logic on this PR checks whether HEAD can be found on remote. If it was logically ORed with the previous logic it would cover any unpushed branches OR unpushed detached HEAD.

I think that covers it, and I can add the test we discussed to supplement the current one.

How does that sound? @aciidb0mb3r

I presume you don't want detached commits that aren't HEAD? If you do then I'm afraid I'm missing the use case...